File: interpreter/pa5c/main.ml
(* pa5c
 * Daniel Coo (djc4ku)
 * Raymond Zhao (rfz5nt)
 *)

(* Won't be considering line number locations, but we have to *)
(* New, Dispatch, Variable, Assign, Integer, Plus *)
open Printf

type exp =
    | New of string (* new Point *)
    | Dispatch of exp * string * (exp list) (* self.foo(a1, a2, ..., an) *)
    | Variable of string (* x *)
    | Assign of string * exp (* x <- 2 + 2 *)
    | Integer of Int32.t (* Int32.t these are 32 bit integer in Cool. can't be 64 bit *)
    | Plus of exp * exp (* x + 5 *)

(* You would read in the annotated AST to get these .... have to to this on our
 * own, not in video *)


(* Environment, Store, Values *)
type cool_address = int
type cool_value =
    | Cool_Int of Int32.t (* Int32.t *)
    | Cool_Bool of bool
    | Cool_String of string
    | Cool_Object of string * ((string * cool_address) list)
    | Void

(* Goal of interpreter is to boil down Expressions into values and to update the
 * store *)

(* Want a function that takes in expressions and produces Cool values,
 * e.g. 2 + 2 = 4 *)

(* Environment: Maps Variable Names -> Cool Addresses *)
(* Could use Hashtables, Maps, ... *)

(* In Ocaml, will use an association list 
 * If x lives at 33 and y lives at address 7, we would have
 * [ ("x", 33)  ; ("y", 7) ]
 *)

type environment = (string * cool_address) list

(* Store: Maps Cool Addresses -> Cool Values *)
type store = (cool_address * cool_value) list

(* Class Map, Implementation Map *)
(* The Class Map maps "Class Names" -> "Attribute names and attribute
 * initializers"
 * TODO: I am not doing types. *)
type class_map = (string * ((string * exp) list)) list

(* Implementation map:
    * Maps ("Class Name", "Method Name") to
    * the method formal parameter names
    * the method body
    * TODO: check what the full version should be! *)
type imp_map =
    (* Class name, method name -> formals and method body *)
    ( (string * string)
      *
      ((string list) * exp) ) list

(************************************************************************************)
(* Debugging and Tracing                                                            *)
(************************************************************************************)

(* NEED TO READ IN EVERY PART OF THE ANNOTATED AST *)


(* Sometimes want to run interpreter with a lot of debugging and sometimes dont
 * *)

(* Debug flag and simple print function *)
(* Set debug to false if you don't want to debug *)
let do_debug = ref true
let debug fmt =
    let handle result_string =
        if !do_debug then printf "%s" result_string
    in
    (* Continuation printf (K) *)
    kprintf handle fmt

(* debug method handles conditional debugging, but it's also helpful to have a
 * way to take all the data types and turn them into strings. Python and Ruby
 * have methods for this!! *)
let rec exp_to_str e =
    match e with
    | New(s) -> sprintf "New(%s)" s
    | Dispatch(ro, fname, args) ->
        (* Fold left, start with empty string and go over all of the arguments*)
        let arg_str = List.fold_left (fun acc elt ->
            acc ^ ", " ^ (exp_to_str elt)
        ) "" args in
        sprintf "Dispatch(%s,%s,[%s])" (exp_to_str ro) fname arg_str
    | Variable(x) -> sprintf "Variable(%s)" x
    | Assign(x,e) -> sprintf "Assign(%s,%s)" x (exp_to_str e)
    | Integer(i) -> sprintf "Integer(%ld)" i
    | Plus(e1,e2) -> sprintf "Plus(%s,%s)" (exp_to_str e1) (exp_to_str e2)


(* Want a way to convert Cool VALUES to strings *)
(* Well.. *)
let value_to_str v =
    match v with
    | Cool_Int(i) -> sprintf "Int(%ld)" i
    | Cool_Bool(b) -> sprintf "Bool(%b)" b
    | Cool_String(s) -> sprintf "String(%s)" s
    | Void -> sprintf "Void"
    | Cool_Object(cname, attrs) -> 
        let attr_str = List.fold_left (fun acc (aname, aaddr) ->
            sprintf "%s, %s=%d" acc aname aaddr
        ) "" attrs in
        sprintf "%s([%s])" cname attr_str

(* Want to print out Environments and Stores *)
let env_to_str env =
    let binding_str = List.fold_left (fun acc (aname, aaddr) ->
        sprintf "%s, %s=%d" acc aname aaddr
    ) "" env in
    sprintf "[%s]" binding_str

(* Print Stores *)
let store_to_str env =
    let binding_str = List.fold_left (fun acc (addr, cvalue) ->
    sprintf "%s, %d=%s" acc addr (value_to_str cvalue)
    ) "" env in
    sprintf "[%s]" binding_str


(************************************************************************************)
(* Evaluation (Interpretation)                                                      *)
(************************************************************************************)

(* Want an evaluation procedure that takes 'so' and 's' 'e' and the expression and
 * returns a new value and a new store *)
(* Read off the rules from Operational Semantics stuff *)
(* Everytime you see a turnstile, convert it into a recursive call *)
let rec eval (so : cool_value)  (* self object *)
             (s : store)        (* store = memory maps addresses to values *)
             (e : environment)  (* maps variables to addresses *)
             (exp : exp)        (* the expression to evaluate *)
             :
             (cool_value *      (* result value *)
             store)             (* updated store *)
             =
    let new_value, new_store = match exp with
    | Integer(i) -> Cool_Int(i), s
    | Plus(e1,e2) -> 
        let v1, s2 = eval so s e e1 in
        let v2, s3 = eval so s2 e e2 in
        let result_value = match v1, v2 with
            | Cool_Int(i1), Cool_Int(i2) ->
                Cool_Int(Int32.add i1 i2)
            | _,_ -> failwith "impossible in plus"
        in
        result_value, s3
    | _ -> failwith "unhandled so far"
    in
    new_value, new_store
    

let main () = begin
    let my_exp = Plus(Integer(51), Integer(31)) in
    debug "my_exp = %s\n" (exp_to_str my_exp);
end ;;
main () ;;

File: interpreter/pa5c/main.py
# Raymond Zhao (rfz5nt)
# Daniel Coo (djc4ku)

# Read in .cl-type file
# Separate into class_map, imp_map, parent_map, and ast
import sys

class Exp:
	def __init__(self, loc=None, exp_kind=None, exp=None):
		self.loc = loc
		self.exp_kind = exp_kind
		self.exp = exp

	def __repr__(self):
		if self.exp_kind == "isvoid":
			return "IsVoid(%s)" % (str(self.exp))
		elif self.exp_kind == "lt":
			return "Lt(%s)" % (str(self.exp))
		elif self.exp_kind == "le":
			return "Le(%s)" % (str(self.exp))
		elif self.exp_kind == "negate":
			return "Negate(%s)" % (str(self.exp))
		elif self.exp_kind == "eq":
			return "Eq(%s)" % (str(self.exp))
		elif self.exp_kind == "times":
			return "Times(%s)" % (str(self.exp))
		elif self.exp_kind == "divide":
			return "Divide(%s)" % (str(self.exp))
		elif self.exp_kind == "minus":
			return "Minus(%s)" % (str(self.exp))
		elif self.exp_kind == "plus":
			return "Plus(%s)" % (str(self.exp))
		elif self.exp_kind == "not":
			return "Not(%s)" % (str(self.exp))
		elif self.exp_kind == "integer":
			return "Integer(%s)" % (str(self.exp))
		elif self.exp_kind == "string":
			return "String(%s)" % (str(self.exp))
		elif self.exp_kind == "true":
			return "Bool(%s)" % (str(self.exp))
		elif self.exp_kind == "false":
			return "Bool(%s)" % (str(self.exp))
		elif self.exp_kind == "identifier":
			return "ID(%s)" % (str(self.exp))
		elif self.exp_kind == "Object":
			return "Internal_Object(%s)" % (str(self.exp))
		elif self.exp_kind == "Int":
			return "Internal_Int(%s)" % (str(self.exp))
		elif self.exp_kind == "IO":
			return "Internal_IO(%s)" % (str(self.exp))
		elif self.exp_kind == "SELF_TYPE":
			return "Internal_SELF_TYPE(%s)" % (str(self.exp))
		elif self.exp_kind == "String":
			return "Internal_String(%s)" % (str(self.exp))
		else:
			return "exp not handled in to string"

#TODO: do for every other expression that is a bitch
class Assign(Exp):
	def __init__(self, loc=None, var=None, exp=None):
		Exp.__init__(self,loc,"assign",exp)
		self.var = var
	def __repr__(self):
		return "Assign(%s,%s)" % (self.var,self.exp)

class Self_Dispatch(Exp):
	def __init__(self, loc=None, fname=None, exp=None):
		Exp.__init__(self,loc,"self_dispatch",exp) 
		self.fname = fname
	def __repr__(self):
		return "Self_Dispatch(%s,%s)" % (self.fname,self.exp)
		
class CoolValue:
	def __init__(self, value_type=None, value=None):
		self.value_type = value_type
		self.value = value
		# TODO: Default values: String = "", Int = 0, Bool = false
class CoolInt(CoolValue):
	def __init__(self, value=0):
		CoolValue.__init__(self,"Int", value)
	def __repr__(self):
		return "CoolInt(%s)" % (self.value)

class CoolString(CoolValue):
	def __init__(self, value="", length=0):
		CoolValue.__init__(self,"String",value)
		self.length = length
	def __repr__(self):
		return "CoolString(%s)" % (self.value)

class CoolBool(CoolValue):
	def __init__(self, value="false"):
		CoolValue.__init__(self,"Bool",value)
	def __repr__(self):
		return "CoolBool(%s)" % (self.value)

class CoolObject:
	def __init__(self, cname=None, attr_and_locs={}):
		self.cname = cname
		self.attr_and_locs = attr_and_locs
	
#TODO: Cool_Object and Void

# Helper functions
def print_list(a):
	for elt in a:
		print elt

def is_int(n):
	try:
		int(n)
		return True
	except ValueError:
		return False

# Debugging and Tracing
do_debug = True
global indent_count
indent_count = 0
def debug_indent():
	global indent_count
	if do_debug:
		for i in range(indent_count):
			print " ",

def debug(e):
	print "%s" % (e)

def print_map(hmap):
	for k in hmap:
		for v in hmap[k]:
			print "%s -> %s" % (k, v)
		print

fname = sys.argv[1]

# Read in .cl-type file
type_file = []
with open(fname, 'r') as fin:
	for line in fin:
		type_file.append(line.rstrip())

io_cmap = []
io_imap = []
io_pmap = []
#TODO: Do we find this useful for any case? LUB?
io_ast  = []

in_cmap = True
in_imap = False
in_pmap = False

for line in type_file:
	if line == "implementation_map":
		in_cmap = False
		in_imap = True
	elif line == "parent_map":
		in_imap = False
		in_pmap = True
	
	if in_cmap:
		io_cmap.append(line)
	elif in_imap:
		io_imap.append(line)
	elif in_pmap:
		io_pmap.append(line)


# Separate ast from pmap
# Start of ast will always be the second integer
int_count = 0
split_pos = 0
for inx, line in enumerate(io_pmap):
	if is_int(line):
		int_count += 1  
	if int_count > 1:
		split_pos = inx
		break

io_ast = io_pmap[split_pos:]
io_pmap = io_pmap[0:split_pos]

# Deerialize the class_map
class_map = {}
imp_map = {}
# li : remaining part of class-map
# helper : function
def read_exp_list(func,num):
	ret_list = []
	for i in range(num):
		ret_list.append(func)
	return ret_list


def read_id(e):
	idloc = e.pop(0)
	idname = e.pop(0)
	t = Exp(idloc, "identifier", idname)
	return t

def read_internal_exp(e):
	loc = e.pop(0)  # loc is always 0
	exp_kind = e.pop(0)
	internal = e.pop(0)
	exp_body = e.pop(0)	# e.g. Object.abort  IO.in_int. See PA4 AST third bullet
	t = Exp(loc, exp_kind, exp_body)
	return t

def read_exp(e):
	# Know that we need to read an exp
	# return a tuple (loc, exp_kind, exp subparts)
	# recurse on the subparts

	# Test if annotated expression. Types are always capitalized
	if e[1][0].isupper():
		loc = e.pop(0)
		exp_type = e.pop(0)
		exp_kind = e.pop(0)
	else:	
		loc = e.pop(0)
		exp_kind = e.pop(0)

	# Read expressions based on exp_kind
	if exp_kind == "assign":
		var = read_id(e)
		t = Assign(loc,var.exp,read_exp(e))
		return t
	elif exp_kind == "new":
		return read_id(e) 
	elif exp_kind == "self_dispatch":
	# WE MIGHT NEED THE METHOD IDENTIFIER
		funcid = read_id(e)
		fname = funcid.exp
		num_of_args = int(e.pop(0))
		t = Self_Dispatch(loc, fname, read_exp_list(read_exp(e),num_of_args))
		return t
	elif exp_kind == "isvoid":
		t = Exp(loc, exp_kind, read_exp(e)) 
		return t
	elif exp_kind == "negate":
		t = Exp(loc, exp_kind, read_exp(e))
		return t
	elif exp_kind in ["plus", "minus", "times", "divide", "lt", "le", "eq"]:
		first_exp = read_exp(e)
		second_exp = read_exp(e)
		t = Exp(loc, exp_kind, [first_exp, second_exp])
		return t
	elif exp_kind == "not":
		t = Exp(loc, exp_kind, read_exp(e))
		return t
	elif exp_kind == "integer":
		int_constant = e.pop(0)
		t = Exp(loc, exp_kind, int_constant)
		return t
	elif exp_kind == "string":
		str_constant = e.pop(0) 
		t = Exp(loc, exp_kind, str_constant)
		return t
	elif exp_kind == "true":
		t = Exp(loc, exp_kind, "true")
		return t
	elif exp_kind == "false":
		t = Exp(loc, exp_kind, "false")
		return t
	elif exp_kind == "identifier":
		return read_id(e)		
	else:
		print "Expression %s not handled in read_exp(e)" % (exp_kind)
	

def read_cmap(cmap_list):
	num_classes = int(cmap_list.pop(0))
	while num_classes > 0:
		# 0 is just for testing. Remove later
		try:
			attrs = []
			cname = cmap_list.pop(0)
			num_attrs = int(cmap_list.pop(0))
			while num_attrs > 0:
				initialize = cmap_list.pop(0)
				attr_name = cmap_list.pop(0)
				attr_type = cmap_list.pop(0)
				attr_exp = []
				if initialize == "initializer":
					attr_exp.append(read_exp(cmap_list))
				attrs.append((attr_name, attr_type, attr_exp))  
				num_attrs -= 1
			num_classes -= 1
			class_map[cname] = attrs
		except ValueError:
			print "ValueError, messed up somewhere in read_cmap"

# key: (class name, method name) value: (formals list, method body)
def read_impmap(imap):
	num_classes = int(imap.pop(0))
	while num_classes > 0:
		try:
			class_name = imap.pop(0)
			num_of_methods = int(imap.pop(0))
			for i in range(num_of_methods):
				method_name = imap.pop(0)
				num_of_formals = int(imap.pop(0))
				formal_list = [imap.pop(0) for i in range(num_of_formals)]	
				parent_class = imap.pop(0)
				
				# Check if internal method
				# Contents of imap if internal
				# 0				imap[0]
				# Object		imap[1]
				# internal		imap[2]
				# Object.abort	imap[3]
				if imap[2] == "internal":
					body_exp = read_internal_exp(imap)
				else:
					body_exp = read_exp(imap)

				imp_map[(class_name,method_name)] = (formal_list,body_exp)
			num_classes -= 1
		except ValueError:
			print "ValueError, messed up while reading the lines from imp map"

read_cmap(io_cmap[1:])
print "CLASS_MAP"
print_map(class_map)

print "IMP_MAP"
read_impmap(io_imap[1:])
print_map(imp_map)



new_location_counter = 1000
def newloc():
	new_location_counter += 1	
	return new_location_counter

# Parameters:
# 	so		: self object
# 	store 	: store maps addresses to values
#	env		: environment maps variables to addresses
#	e		: the expression to evaluate
#
# Return Value:
#	(new_value, updated_store)

main_class = class_map['Main']
main_imp = imp_map[('Main','main')]
my_exp = main_imp[1]
print "my_exp: %s" % (my_exp)

def eval(self_object,store,environment,exp):
	global indent_count
	indent_count += 2
	debug_indent() ; debug("eval: %s" % (exp))
	debug_indent() ; debug("so    = %s" % (self_object))
	debug_indent() ; debug("store = %s" % (store))
	debug_indent() ; debug("env   = %s" % (environment))


	if exp.exp_kind == "assign":
	# loc, var, exp
		(v1,s2) = eval(self_object,store,environment,exp.exp)	
		l1 = environment[exp.var]	
		del s2[l1] 
		s3 = s2
		s3[l1] = v1
		debug_indent() ; debug("ret = %s\n" % (v1))
		debug_indent() ; debug("rets = %s\n" % (s3))
		indent_count -= 2
		return (v1,s3)
	elif exp.exp_kind == "new":
		cname = exp.exp
		attrs_and_inits = class_map[cname]
		new_attrs_locs = [newloc() for x in attrs_and_inits]

			
			
		# need assign
		# get class
			
		pass
	elif exp.exp_kind == "self_dispatch":
	# self_dispatch = loc, exp_kind, fname, exp = [args] 
		# Evaluate arguments IN ORDER
		current_store = store
		arg_values = []
		for arg in exp.exp:
			(new_value, new_store) = eval(self_object,store,environment,arg)
			current_store = new_store	
			arg_values.append(new_value)	

		# Evaluate Receiver Object (for self, it's just the self_object)
		
		(v0,s_nplus2) = eval(self_object,current_store,environment,e0)

		# Look up things in implementation map
		# v0 should be a CoolObject with the name of class and attrs_and_locs			
		# TODO: if it's not in there what happens? WELLLLLLLLL	
		(formals, body) = imp_map[v0[0],fname]
		# TODO: Incomplete, working on new

		pass
	elif exp.exp_kind == "isvoid":
		pass
	elif exp.exp_kind == "negate":
		pass
	elif exp.exp_kind == "plus": #need other operations
		# Get each integer from plus expression
		e1 = my_exp.exp[0]
		e2 = my_exp.exp[1]
		print "e1: %s" % (e1)
		print "e2: %s" % (e2)
		v1, s2 = eval(self_object,store,environment,e1)
		v2, s3 = eval(self_object,store,environment,e2)
		new_value = v1.value + v2.value
		debug_indent() ; debug("ret = %s" % (new_value))
		debug_indent() ; debug("rets = %s" % (store))
		indent_count -= 2
		return (CoolInt(new_value), store)
	elif exp.exp_kind == "not":
		pass
	elif exp.exp_kind == "integer":
		value = int(exp.exp)
		debug_indent() ; debug("ret = %s" % (value))
		debug_indent() ; debug("rets = %s" % (store))
		indent_count -= 2
		return (CoolInt(value), store)
	elif exp.exp_kind == "string":
		value = str(exp.exp)
		length = len(value)
		debug_indent() ; debug("ret = %s" % (value))
		debug_indent() ; debug("rets = %s" % (store))
		indent_count -= 2
		return (CoolString(value,length), store)
	elif exp.exp_kind == "true":
		pass
	elif exp.exp_kind == "false":
		pass
	elif exp.exp_kind == "identifier":
		pass
	else:
		print "Expression %s not handled" % (exp.exp_kind)

# ----------MAIN----------
	
# Environment, Store, and Values
# Environment is a list of tuples
# e.g. x lives at address 33 and y lives address 7
# [ ('x', 33), ('y', 7) ]
env = {}
# Store is a dictionary that maps addresses to their values
# e.g. 
store = {}
# Self Object
self_object = []
(new_value, new_store) = eval(self_object, store, env, my_exp)
print new_value.value

# self would be nothing at this point
#eval self store env (new Main).main() 
#new Main becomes the self object










